### ISSUE TITLE
Captures the wrong line on streaming video

### ISSUE THREAD

**Issue by kanjieater** [asbplayer/asbplayer#330]
I've tried out a couple different mining methods recently, and while asbplayer seems like the smoothest workflow there's a part of the flow that seems very error prone compared to other methods (anacreon's or tatsumoto's mpv scripts).

In all 3 methods, you can add a card with yomichan, then have the tool add media to the last created card. In both of the other tools they automatically match the sentence in the anki card with the sentence in the subtitles. 

It seems like (unless I'm missing something), asbplayer doesn't do this. It will just add whatever audio you last had when you hit the hotkey. I'm trying to use the `Update last-created Anki card with asbplayer-captured media. ` hotkey to consistently add audio and screenshots to the cards I'm creating. 

If I want to add a word for a past subtitle do I have to go back to get asbplayer to get the audio and screenshot for it? If I unpause the video then hit the hotkey and the subtitle moves can asbplayer detect the audio on the card won't match the subtitle text, and add the right one. If I have the subtitle window selected and hit the hotkey to add image & audio, it says added, but doesn't add anything (I have to select the player window and try again).

tldr;
I want to only add the audio & screenshot from the matching subtitle, and it seems like switching between windows (subs & video windows) makes this more error prone in asb than other solutions.

Separate thought:
Ideally asbplayer browser ext, yomichan or ankiconnect would just hook into asbplayer and eliminate the need for an asbplayer hotkey to capture media. It would be super smooth if as the cards were being created via ankiconnect, all screenshot & audio was captured automatically.

**Comment by killergerbah** [comment]
> It seems like (unless I'm missing something), asbplayer doesn't do this. It will just add whatever audio you last had when you hit the hotkey. 

Yeah that's right. asbplayer just adds the media for the subtitle currently being displayed on the video player side.

> If I want to add a word for a past subtitle do I have to go back to get asbplayer to get the audio and screenshot for it?

Right now, that is the case unfortunately.

> If I have the subtitle window selected and hit the hotkey to add image & audio, it says added, but doesn't add anything (I have to select the player window and try again).

This is a known issue with AnkiConnect (https://github.com/FooSoft/anki-connect/issues/82) that I need to implement the workaround for.

> Ideally asbplayer browser ext, yomichan or ankiconnect would just hook into asbplayer and eliminate the need for an asbplayer hotkey to capture media. It would be super smooth if as the cards were being created via ankiconnect, all screenshot & audio was captured automatically.

Without implementing any asbplayer-specific integrations on the Yomichan or AnkiConnect side, it would be possible for asbplayer to constantly poll AnkiConnect for new cards. Once a new card is added, asbplayer could find the matching subtitle and automatically capture media + update that card. Thanks for the cool idea.

As for updating the last card with the media matching the sentence in that card, I believe this would be an overall improvement over the current behavior. My only concern is whether that behavior would be astonishing to some people, considering the other mining behaviors ("open dialog" or "do nothing") are always for the currently-displaying subtitle. I'll give this more thought; maybe this new behavior could be a new setting, or maybe the automatic-update idea above could be an alternative solution.

**Comment by killergerbah** [comment]
> It seems like the MPV solutions can just clip audio/video in the background without requiring user interaction. Is there some browser-side way to do this that asbplayer could do in the background? If not I'm guessing you'd have to use some sort of "MPV server" to do the clipping, which seems doable, but obviously a significant change from the current flow.

In the case of local video, it would be possible and there is an issue to do this. The "MPV server" you're referring to would basically be a copy of ffmpeg. This is how animebook (https://github.com/animebook/animebook.github.io) does it.

In the case of streaming video, the most reliable way to capture media is using the tab-based recording/capturing APIs, since these bypass any security implemented by the website itself. As a result asbplayer has to seek to the desired point in the video to record audio.
